finance_dl/discover.py

```diff
@@ -0,0 +1,176 @@
+            os.makedirs(self.output_directory)
+
+    def check_after_wait(self):
+        check_url(self.driver.current_url)
+
+    def find_account_last4(self):
+        return self.driver.find_element_by_xpath(XPATH_OF_LAST_FOUR_DIGITS)
+
+    def login(self):
+        elements = self.driver.find_elements_by_xpath(XPATH_OF_LAST_FOUR_DIGITS)
+        if not elements:
+            logger.info('Initiating log in')
+            url = "https://portal.discover.com/customersvcs/universalLogin/ac_main"
+            self.driver.get(url)
+
+            (username, password), = self.wait_and_return(
+                self.find_username_and_password_in_any_frame)
+            logger.info('Entering username and password')
+            username.clear()
+            password.clear()
+            username.send_keys(self.credentials['username'])
+            password.send_keys(self.credentials['password'])
+            with self.wait_for_page_load(timeout=30):
+                password.send_keys(Keys.ENTER)
+            self.check_after_wait()
+            logger.info('Logged in')
+        else:
+            logger.info("Already logged in")
+        try:
+            account = self.wait_and_return(self.find_account_last4, timeout=7)[0]
+            account = account.text
+            if len(account) != 4:
+                raise Exception("{}: Either identified tag is not the account number, or they've changed its format.".format(account))
+            self.account_name = account
+        except Exception as e:
+            logger.exception(str(e))
+            logger.exception("Failed to find account tag after attempting to log in")
+            raise NoSuchElementException()
+
+    def download_ofx(self):
+        # Downloads all transactions in current calendar year
+        this_year = datetime.date.today().year
+        start_date = datetime.datetime.strptime(str(this_year), '%Y')
+        end_date = datetime.datetime.strptime(str(this_year + 1), '%Y')
+        logging.info('Fetching history for all of {}'.format(this_year))
+
+        stdt = start_date.strftime('%Y%m%d')
+        enddt = end_date.strftime('%Y%m%d')
+        fname = "Discover-{}.ofx".format(this_year)
+        # Example URL:
+        # https://card.discover.com/cardmembersvcs/ofxdl/ofxWebDownload?stmtKey=W&startDate=20191014&endDate=20191014&fileType=QFX&bid=9625&fileName=Discover-RecentActivity-20191014.qfx
+        qfx_url = "https://card.discover.com/cardmembersvcs/ofxdl/ofxWebDownload?stmtKey=W&startDate={}&endDate={}&fileType=QFX&bid=9625&fileName={}"
+        qfx_url = qfx_url.format(stdt, enddt, fname)
+        logging.info("Downloading from URL: {}".format(qfx_url))
+        # this is necessary because otherwise it locks on the get() for a while
+        self.driver.set_page_load_timeout(5)
+        try:
+            self.driver.get(qfx_url)
+        except TimeoutException:
+            pass
+        src = os.path.join(self.download_dir, fname)
```

Comment:
You may find it simpler to use `self.get_downloaded_file()`

Reply:
(In conjunction with `self.wait_and_return`, you can refer to other sources for details.)

Reply:
Thanks. I hadn't noticed that routine.

Reply:
`get_downloaded_file` does make the code cleaner, but I don't see how to solve the real issue: that the `get()` call blocks. The other scapers based on scraper_lib all do a `link.click()` before `wait_and_return(self.get_downloaded_file)`, whereas I know the URL I need to access and don't interact with the HTML objects.
Do you see a way around?